int32 num_points
---